### PR TITLE
Refactor sdk Dockerfile templates

### PR DESCRIPTION
**Pull request by mthalman**
These changes refactor the sdk Dockerfile templates to enable more sharing of content. All the templates have been distilled down to a Linux and Windows primary template that can be used across .NET versions and OS types.

There was some refactoring in the other existing templates as well to account for a common implementation of certain patterns. Notably, the `eng/dockerfile-templates/Dockerfile.linux.download-and-install` template was added which encapsulates the downloading and installation of either tarballs or RPMs.

### CODE REVIEW

File: eng/dockerfile-templates/sdk/Dockerfile.windows [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+
+    _ SINGLE STAGE
+
+}}FROM {{aspnetBaseTag}}
+
+{{InsertTemplate("Dockerfile.envs")}}
+
+{{InsertTemplate("Dockerfile.windows.install-components")}}
+
+{{InsertTemplate("Dockerfile.windows.set-path")}}
```
Comment (MichaelSimons) [inline eng/dockerfile-templates/sdk/Dockerfile.windows:18]: I think renaming this in terms of pwsh would help the readability.  It helps clarify that is not adding dotnet to the path.

File: eng/dockerfile-templates/sdk/Dockerfile.linux.install-sdk [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+        based-on-buildpack-deps: Indicates whether the associated Dockerfile is based on buildpack-deps
+        use-local-version-var: Whether to define a local variable for the .NET SDK version instead of
+            referencing the environment variable. ^
+
+    set dotnetVersion to join(slice(split(PRODUCT_VERSION, "."), 0, 2), ".") ^
+    set isAlpine to find(OS_VERSION, "alpine") >= 0 ^
+    set isFullMariner to defined(match(OS_VERSION, "^cbl-mariner\d+\.\d+$")) ^
+    set platform to when(isAlpine, "linux-musl", "linux") ^
+    set dotnetSuffix to when(dotnetVersion = "3.1", " Core", "") ^
+    set tarballFile to "dotnet.tar.gz" ^
```
Comment (MichaelSimons) [inline eng/dockerfile-templates/sdk/Dockerfile.linux.install-sdk:12]: I'm not finding how this is used.

File: eng/dockerfile-templates/sdk/Dockerfile.linux.install-sdk [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+        "/Sdk/$dotnet_sdk_version/dotnet-sdk-$dotnet_sdk_version-", platform, "-", ARCH_SHORT, ".tar.gz") ^
+    set sdkVersionVar to when(ARGS["use-local-version-var"], "$sdk_version", "$DOTNET_SDK_VERSION") ^
+    set runtimeVersionVar to when(ARGS["use-local-version-var"], "$dotnet_version", "$DOTNET_VERSION") ^
+    set baseUrl to VARIABLES[cat("base-url|", dotnetVersion, "|", VARIABLES["branch"])] ^
+    set rpms to [
+        [
+            "filename": "dotnet.rpm",
+            "url": cat(baseUrl, "/Sdk/", sdkVersionVar, "/dotnet-sdk-", sdkVersionVar, "-", ARCH_SHORT, ".rpm"),
+            "sha": VARIABLES[join(["sdk", dotnetVersion, "linux-rpm", ARCH_SHORT, "sha"], "|")],
+            "sha-var-name": "dotnet_sha512"
```
Comment (MichaelSimons) [inline eng/dockerfile-templates/sdk/Dockerfile.linux.install-sdk:25]: "dotnet_sha512" is referenced numerous times.  Consider encapsulating in a shared variable.

**Review by MichaelSimons — APPROVED** [review]

File: eng/dockerfile-templates/sdk/Dockerfile.linux.install-sdk (hunk 2 above)
Comment (mthalman) [reply eng/dockerfile-templates/sdk/Dockerfile.linux.install-sdk]: Leftover after refactoring. Removed.

File: eng/dockerfile-templates/sdk/Dockerfile.windows (hunk 1 above)
Comment (mthalman) [reply eng/dockerfile-templates/sdk/Dockerfile.windows]: Renamed to Dockerfile.windows.set-powershell-path.

**Review by MichaelSimons — APPROVED** [review]

**Pull request merged**